### ISSUE TITLE
Hyperparameters for reproducibility

### ISSUE THREAD

**Issue by CubicQubit** [shchur/ifl-tpp#16]
Thank you for providing your code, it is very well organized.

Would it also be possible to release the hyperparameters used for each of the datasets in the paper? This comes from me having trouble reproducing the set of numbers in Table 3 in the appendix for the provided datasets (StackOverflow and other synthetic datasets)? For example, running the code out of the box gives:

Breaking due to early stopping at epoch 172
Negative log-likelihood:
 - Train: 1151.5
 - Val:   1179.1
 - Test:  1131.5

For stack overflow dataset at seed = 0, which is not the 14.4 number.

In addition, I was wondering if it would be possible to release the code for event time prediction using history?

**Comment by shchur** [comment]
Hi, just to double check, are you using the code on the branch [`original-code`](https://github.com/shchur/ifl-tpp/tree/original-code), as described in the [readme](https://github.com/shchur/ifl-tpp#refactored-code)?

There are a few changes between the original code used to run the experiments in the paper and the code currently on the master branch, the main difference being how we normalize the NLL (by # of events on `original-code` vs. by the interval length on `master`). This is the most likely reason for such a big difference in the results. As far as I remember, the results were pretty robust to the choice of the hyperparameters, so I don't think they would make such a big difference. You can see the details about the hyperparameters in the appendix of the paper.

**Comment by won-bae** [comment]
Thanks for the clarification. 
Actually could you elaborate, what you meant by normalizing NLL using the interval length on master? I observe the summation and it seems like the current implementation in master is right according to math. But I don't observe this normalization part. Could you specify where it is implemented? Thank you.

**Comment by shchur** [comment]
Oops, I guess the description in my comment doesn't match the implementation in the code. To normalize by the interval length, we would need to modify https://github.com/shchur/ifl-tpp/blob/e7ebab1ceab56cee440bd8e99b5c1bd42d6ada07/code/dpp/models/recurrent_tpp.py#L137
to
```python
        return (log_p.sum(-1) + log_surv_last) / batch.inter_times.sum(-1)  # (batch_size,)
```

**Comment by won-bae** [comment]
Hi, thank you for correcting it!
Unfortunately, if you run the revised version, since  `batch.inter_times.sum(-1)` is very large compared to `(log_p.sum(-1) + log_surv_last)`, the loss becomes almost 0, which is not what we want I guess. Could you look into it once again?
In fact, I do not quite understand why this normalization is necessary. If my understanding is correct, `log_p.sum(-1) + log_surv_last` corresponds to the log of joint probability of a sequence tau_i e.g., log P(tau_i). Normalized NLL would not change training much (as long as learning rate is adjusted accordingly) but it does not seem aligning with math. Is there any particular reason to normalize it? Thank you

**Comment by shchur** [comment]
Apologies for such a late reply, the notification got lost in my inbox.

One reason for normalizing by time is when we deal with sequences observed on time intervals of different length [0, T] (i.e., $T$ is different for each sequence). Suppose we had 100 sequences generated from the same TPP (e.g., same Poisson process), but the value of $T$ was different for each sequence, and we train the model using minibatches of size 10. If we normalize the NLL for each sequence by $1/T$, the loss will on average be the same for any batch. If we don't normalize, the loss will be very different across sequences with different values of $T$, even though they were generated by the same model. 

As you correctly pointed out, normalization won't really affect the optimization, given that we tune the learning rate appropriately.

I guess for the dataset that you're considering the time is measured in seconds, and therefore the value of $T$ is too big and the normalized NLL gets close to zero. One solution would be to rescale the times (e.g., measured in hours). 

To summarize, the choice of whether we normalize the NLL by time is not that essential and it's perfectly fine not to normalize it.

**Comment by won-bae** [comment]
I see. Thank you so much for the clarification and suggestion!
